### PR TITLE
Added forcecephkernelclient as startup parameter to force enabling ceph

### PR DESCRIPTION
**Pull request by shaas**
Added "forcecephkernelclient" as parameter to force the use of the ceph kernel client for kernels < 4.17

This is need as SUSE Enterprise Storage 6 does not support fuse but ceph quotas got back-ported to kernels < 4.17

Signed-off-by: Stefan Haas <shaas@suse.com>

### CODE REVIEW

**Comment by humblec** [comment]
>This is need as SUSE Enterprise Storage 6 does not support fuse but ceph quotas got back-ported to kernels < 4.17
>

@shaas are you saying SUSE kernel did a backport of quota support for kernels < 4.17 ?

**Comment by humblec** [comment]
We have some concerns on enabling kernel client without quota support as mentioned in https://github.com/ceph/ceph-csi/issues/617. If SUSE kernel support it, I would rather go for suse specific kernel check and enablement than a general one.

**Comment by BlaineEXE** [comment]
#617 is a good instance of a user wanting this feature. I see a lot of good arguments here for why a hard limit on 4.17 is a poor choice, like here: https://github.com/ceph/ceph-csi/pull/623#issuecomment-533142405

**Comment by shaas** [comment]
> can you also update the doc for this new variable here https://github.com/ceph/ceph-csi/blob/master/docs/deploy-cephfs.md#configuration

Done

**Comment by Madhu-1** [comment]
>7.401µs, max_per_file_from_linter: 3.717µs, diff: 3.568µs, exclude-rules: 3.031µs, skip_files: 1.471µs 
pkg/cephfs/volumemounter.go:73:6: S1002: should omit comparison to bool constant, can be simplified to `conf.ForceKernelClient` (gosimple)
		if conf.ForceKernelClient == true || majorVers >= 4 && minorVers >= 17 {
		   ^
@shaas we have CI failure can you please fix it, Thanks

**Comment by shaas** [comment]
> > 7.401µs, max_per_file_from_linter: 3.717µs, diff: 3.568µs, exclude-rules: 3.031µs, skip_files: 1.471µs
> > pkg/cephfs/volumemounter.go:73:6: S1002: should omit comparison to bool constant, can be simplified to `conf.ForceKernelClient` (gosimple)
> > if conf.ForceKernelClient == true || majorVers >= 4 && minorVers >= 17 {
> > ^
> > @shaas we have CI failure can you please fix it, Thanks

Done

File: cmd/cephcsi.go [hunk 1]
```diff
@@ -63,6 +63,7 @@ func init() {
 
 	// cephfs related flags
 	flag.StringVar(&conf.MountCacheDir, "mountcachedir", "", "mount info cache save dir")
+	flag.BoolVar(&conf.ForceKernelClient, "forcecephkernelclient", false, "enable Ceph Kernel clients on kernel < 4.17 which support quotas")
```
Comment (humblec) [inline cmd/cephcsi.go]: which does not support quotas?

File: pkg/util/util.go [hunk 2]
```diff
@@ -77,9 +77,13 @@ type Config struct {
 
 	IsControllerServer bool // if set to true start provisoner server
 	IsNodeServer       bool // if set to true start node server
+
 	// rbd related flag
 	Containerized bool // whether run as containerized
 
+	// cephfs related flags
+	ForceKernelClient bool // force to use the ceph kernel client even if the kernel is < 4.17
```
Comment (humblec) [inline pkg/util/util.go]: Isnt it better to put the fied as "forcekcephfs" ?

File: cmd/cephcsi.go (hunk 1 above)
Comment (krig) [reply cmd/cephcsi.go]: No, the flag is to enable the kernel client on kernels with version <4.17 that do support quotas (via backported patches). All kernels >=4.17 support quotas and don't need the flag

**Comment by Madhu-1** [comment]
@humblec anything else is pending on this one?

**Comment by Madhu-1** [comment]
Fixes: https://github.com/ceph/ceph-csi/pull/664

**Comment by Madhu-1** [comment]
@nixpanic @humblec PTAL

**Comment by humblec** [comment]
@Madhu-1 there were some unanswerd review comments

**Comment by Madhu-1** [comment]
added DNM till we get answers for @humblec question, @shaas can you please take a look, would like to pull  this for 1.2.2 release

**Comment by humblec** [comment]
All,  this was possible in `1.0.0` does not justify much and if we allow kcephfs without quota support we are actually against kubernetes norms where it says the user should get "atleast requested space".  however considering this is a concern from many users, the best bet would be "defaulting to false" which this PR does and adding "extra" documentation in "README" about this functionality with disclaimers in place which says, **this is not recommended/supported if the kernel does not support quota.** 
@shaas we have to document this in README before getting this PR in

File: README.md [hunk 3]
```diff
@@ -77,6 +77,9 @@ NOTE:
 compatible changes in the future, and is thus not recommended
 for production use.
 
+`NOTE` : The parameter `-forcecephkernelclient` enables the Kernel CephFS mounter on kernels < 4.17.
```
Comment (Madhu-1) [inline README.md]: I believe this Note should go to https://github.com/ceph/ceph-csi/blob/master/docs/deploy-cephfs.md which talks about cephfs deployment

**Comment by Madhu-1** [comment]
@shaas static check is failing
```
mdl --style scripts/mdl-style.rb ./docs/deploy-rbd.md 
mdl --style scripts/mdl-style.rb ./docs/deploy-cephfs.md 
./docs/deploy-cephfs.md:65: MD012 Multiple consecutive blank lines
A detailed description of the rules is available at https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md
mdl --style scripts/mdl-style.rb ./docs/coding.md 
mdl --style scripts/mdl-style.rb ./e2e/README.md 
mdl --style scripts/mdl-style.rb ./README.md 
mdl --style scripts/mdl-style.rb ./charts/ceph-csi-cephfs/README.md 
mdl --style scripts/mdl-style.rb ./charts/ceph-csi-rbd/README.md 
Makefile:45: recipe for target 'static-check' failed
make: *** [static-check] Error 123
The command "make static-check" exited with 2.

```

File: pkg/util/util.go (hunk 2 above)
Comment (shaas) [reply pkg/util/util.go]: Changed it

**Comment by shaas** [comment]
> please update the target branch to master, instead of release-v1.2.0

Done

**Comment by Madhu-1** [comment]
@humblec PTAL

**Comment by shaas** [comment]
Is there anything still needed/missing from my side?

**Comment by Madhu-1** [comment]
@humblec @ShyamsundarR @nixpanic PTA

**Comment by humblec** [comment]
@shaas we have to rebase this PR .

**Pull request merged**